### PR TITLE
Download jcc err.log

### PR DESCRIPTION
**Pull request by cjx10**
Enables Google/oss-fuzz#11789

### CODE REVIEW

**Comment by cjx10** [comment]
JOB: https://console.cloud.google.com/kubernetes/job/us-central1-c/llm-experiment/default/ofg-pr-212-jim
REPORT: https://llm-exp.oss-fuzz.com/Result-reports/ofg-pr/2024-04-10-212-jim-comparison/index.html
BUCKET: https://console.cloud.google.com/storage/browser/oss-fuzz-gcb-experiment-run-logs/Result-reports/ofg-pr/2024-04-10-212-jim-comparison
BUCKET GS: `gs://oss-fuzz-gcb-experiment-run-logs/Result-reports/ofg-pr/2024-04-10-212-jim-comparison`

Uploaded error logs located at `results/*/logs/build/*.err.log`

File: experiment/workdir.py [hunk 1]
```diff
@@ -77,5 +77,9 @@ def build_logs_target(self, generated_target_name: str, iteration: int):
     return os.path.join(self.build_logs,
                         f'{generated_target_name}-F{iteration}.log')
 
+  def err_logs_target(self, generated_target_name: str, iteration: int):
```
Comment (DonggeLiu) [inline experiment/workdir.py:80]: nit: Return type hinting?
Sorry, I just realized that other functions don't have that either, which may have confused you.

I can fix them in later PRs, could you please add this?

Comment (cjx10) [reply experiment/workdir.py:80]: Sure, can do the same for the rest of:)

Comment (DonggeLiu) [reply experiment/workdir.py:80]: Probably in a different PR, just to ensure one task per PR : )

**Comment by cjx10** [comment]
JOB: https://console.cloud.google.com/kubernetes/job/us-central1-c/llm-experiment/default/ofg-pr-212-jim
REPORT: https://llm-exp.oss-fuzz.com/Result-reports/ofg-pr/2024-04-16-212-jim-comparison/index.html
BUCKET: https://console.cloud.google.com/storage/browser/oss-fuzz-gcb-experiment-run-logs/Result-reports/ofg-pr/2024-04-16-212-jim-comparison
BUCKET GS: `gs://oss-fuzz-gcb-experiment-run-logs/Result-reports/ofg-pr/2024-04-16-212-jim-comparison`

Uploaded error logs located at `results/*/logs/build/*.err.log`

**Comment by DonggeLiu** [comment]
@cjx10: There seems to be an error with the report, could you please check?
Thanks!

**Comment by cjx10** [comment]
> There seems to be an error with the report

JCC binary was not uploaded properly, now fixed

**Comment by cjx10** [comment]
Fail to upload err.log currently halts the cloud build. There exists projects that don't use jcc to compile the source, e.g. [civetweb](https://github.com/civetweb/civetweb/blob/master/Makefile#L80C3-L80C13). This pr will break these projects before `'allowFailure': True` is added [here](https://github.com/google/oss-fuzz/commit/12041481001bca0f9b2f063ce7cc330ec55a0633#diff-ed7b0455ec9bfd97499697c6be191669e6f378243e24ba14981b4b7f092ffd6cR100)

**Comment by DonggeLiu** [comment]
> Fail to upload err.log currently halts the cloud build. `civetweb` doesn't use jcc to compile the source: https://github.com/civetweb/civetweb/blob/master/Makefile#L80C3-L80C13. This pr will break this projects before `'allowFailure': True` is added [here](https://github.com/google/oss-fuzz/commit/12041481001bca0f9b2f063ce7cc330ec55a0633#diff-ed7b0455ec9bfd97499697c6be191669e6f378243e24ba14981b4b7f092ffd6cR100) Only this project is found to be not using jcc at the moment.

Thanks for reporting this, it is an interesting finding.
Could you please edit the bash command to upload a file to indicate an upload error (e.g., the file content is 'Failed to generate JCC error log.') so that we can find all projects that failed on this with a full experiment?
This will also allow OSS-Fuzz-Gen to capture and address JCC-related failures.

Adding `'allowFailure': True` can ideed address upload failure, but it will hide the error from us.

**Comment by cjx10** [comment]
google/oss-fuzz#11822 prevents failing at upload err.log step

File: experiment/builder_runner.py [hunk 2]
```diff
@@ -437,6 +441,17 @@ def build_and_run(self, generated_project: str, target_path: str,
         print(f'Cannot find cloud build log of {os.path.realpath(target_path)} '
               f':{build_log_name}')
 
+    with open(self.work_dirs.err_logs_target(generated_target_name, iteration),
+              'wb') as f:
+      blob = bucket.blob(err_log_name)
+      if blob.exists():
+        print(f'Downloading cloud build log of {os.path.realpath(target_path)}:'
+              f' {err_log_name} to {f}')
```
Comment (DonggeLiu) [inline experiment/builder_runner.py:449]: Let's use logging.info as a better practice. Thanks!

File: experiment/builder_runner.py [hunk 3]
```diff
@@ -437,6 +441,17 @@ def build_and_run(self, generated_project: str, target_path: str,
...
+        blob.download_to_file(f)
+      else:
+        print(f'Cannot find jcc error log of {os.path.realpath(target_path)} '
+              f':{err_log_name}')
```
Comment (DonggeLiu) [inline experiment/builder_runner.py:453]: logging.warning, thanks : )

**Review by DonggeLiu — APPROVED** [review]
Wow, thanks for fixing them all!

File: experiment/workdir.py [hunk 4]
```diff
@@ -81,6 +81,10 @@ def build_logs_target(self, generated_target_name: str, iteration: int):
     return os.path.join(self.build_logs,
                         f'{generated_target_name}-F{iteration}.log')
 
+  def err_logs_target(self, generated_target_name: str, iteration: int) -> str:
```
[the remark this answers is not in the document]
Comment (cjx10) [reply experiment/workdir.py]: Done

**Comment by DonggeLiu** [comment]
Shall we merge this and https://github.com/google/oss-fuzz/pull/11822, if they have been tested?

**Comment by cjx10** [comment]
> Shall we merge this and [google/oss-fuzz#11822](https://github.com/google/oss-fuzz/pull/11822)

Yes please :)

**Comment by DonggeLiu** [comment]
Has this commit [download jcc err.log](https://github.com/google/oss-fuzz-gen/pull/212/commits/5eee096e69699878de4eccbe586e2e4f69970eb3) been tested?

**Comment by DonggeLiu** [comment]
I'ved merged https://github.com/google/oss-fuzz/pull/11822, could you please update JCC and re-run an exp here if the latest code in this PR is not tested?
Thanks!

**Comment by cjx10** [comment]
> re-run an exp here

JOB: https://console.cloud.google.com/kubernetes/job/us-central1-c/llm-experiment/default/ofg-pr-212-jim
REPORT: https://llm-exp.oss-fuzz.com/Result-reports/ofg-pr/2024-04-23-212-jim-comparison/index.html
BUCKET: https://console.cloud.google.com/storage/browser/oss-fuzz-gcb-experiment-run-logs/Result-reports/ofg-pr/2024-04-23-212-jim-comparison
BUCKET GS: `gs://oss-fuzz-gcb-experiment-run-logs/Result-reports/ofg-pr/2024-04-23-212-jim-comparison`

**Comment by DonggeLiu** [comment]
Thanks!
If the report looks good to you, we can merge this, update https://github.com/google/oss-fuzz-gen/pull/208, and launch a full experiment from there.

**Comment by cjx10** [comment]
> update #208

That pr doesn't use err.log, will experiment including non clang diagnostics using err.log in a new one (mainly those from gnu ld)

**Pull request merged**